Mathlib/RingTheory/AdicCompletion/Basic.lean

```diff
@@ -202,24 +202,57 @@ instance : Neg (AdicCompletion I M) where
 instance : Sub (AdicCompletion I M) where
   sub x y := ⟨x.val - y.val, by simp [x.property, y.property]⟩
 
-instance : SMul ℕ (AdicCompletion I M) where
-  smul n x := ⟨n • x.val, by simp [x.property]⟩
+instance instSMul [SMul S R] [SMul S M] [IsScalarTower S R M] : SMul S (AdicCompletion I M) where
+  smul r x := ⟨r • x.val, by simp [x.property]⟩
+
+@[simp, norm_cast] lemma val_zero : (0 : AdicCompletion I M).val = 0 := rfl
+
+lemma val_zero_apply (n : ℕ) : (0 : AdicCompletion I M).val n = 0 := rfl
+
+variable {I M}
+
+@[simp, norm_cast] lemma val_add (f g : AdicCompletion I M) : (f + g).val = f.val + g.val := rfl
+@[simp, norm_cast] lemma val_sub (f g : AdicCompletion I M) : (f - g).val = f.val - g.val := rfl
+@[simp, norm_cast] lemma val_neg (f : AdicCompletion I M) : (-f).val = -f.val := rfl
 
-instance : SMul ℤ (AdicCompletion I M) where
-  smul n x := ⟨n • x.val, by simp [x.property]⟩
+lemma val_add_apply (f g : AdicCompletion I M) (n : ℕ) : (f + g).val n = f.val n + g.val n := rfl
+lemma val_sub_apply (f g : AdicCompletion I M) (n : ℕ) : (f - g).val n = f.val n - g.val n := rfl
+lemma val_neg_apply (f : AdicCompletion I M) (n : ℕ) : (-f).val n = -f.val n := rfl
+
+/- No `simp` attribute, since it causes `simp` unification timeouts when considering
+the `Module (AdicCompletion I R) (AdicCompletion I M)` instance (see `AdicCompletion/Algebra`). -/
+@[norm_cast]
+lemma val_smul [SMul S R] [SMul S M] [IsScalarTower S R M] (r : S) (f : AdicCompletion I M) :
+    (r • f).val = r • f.val := rfl
+
+lemma val_smul_apply [SMul S R] [SMul S M] [IsScalarTower S R M] (r : S) (f : AdicCompletion I M)
+    (n : ℕ) : (r • f).val n = r • f.val n := rfl
+
+@[ext]
+lemma ext {x y : AdicCompletion I M} (h : ∀ n, x.val n = y.val n) : x = y := Subtype.eq <| funext h
+
+variable (I M)
 
 instance : AddCommGroup (AdicCompletion I M) :=
   let f : AdicCompletion I M → ∀ n, M ⧸ (I ^ n • ⊤ : Submodule R M) := Subtype.val
-  Subtype.val_injective.addCommGroup f rfl (fun _ _ ↦ rfl) (fun _ ↦ rfl) (fun _ _ ↦ rfl)
-    (fun _ _ ↦ rfl) (fun _ _ ↦ rfl)
-
-instance : SMul R (AdicCompletion I M) where
-  smul r x := ⟨r • x.val, by simp [x.property]⟩
+  Subtype.val_injective.addCommGroup f rfl val_add val_neg val_sub (fun _ _ ↦ val_smul ..)
+    (fun _ _ ↦ val_smul ..)
 
-instance : Module R (AdicCompletion I M) :=
+instance [Semiring S] [SMul S R] [Module S M] [IsScalarTower S R M] :
+    Module S (AdicCompletion I M) :=
   let f : AdicCompletion I M →+ ∀ n, M ⧸ (I ^ n • ⊤ : Submodule R M) :=
     { toFun := Subtype.val, map_zero' := rfl, map_add' := fun _ _ ↦ rfl }
-  Subtype.val_injective.module R f (fun _ _ ↦ rfl)
+  Subtype.val_injective.module S f val_smul
+
+instance instIsScalarTower [SMul T S] [SMul T R] [SMul S R] [SMul T M] [SMul S M]
+    [IsScalarTower S R M] [IsScalarTower T S M] [IsScalarTower T R M] :
```

Comment:
nit:
```suggestion
    [IsScalarTower S R M] [IsScalarTower T R M] [IsScalarTower T S M] :
```
so that the "actually needed to prove this" argument is last.